tools/rapids-docker-multiarch-from-s3

```diff
@@ -0,0 +1,101 @@
+#!/bin/bash
+# A utility script that downloads all docker images from S3
+# and assembles them into multiarch tags
+# Optional Positional Arguments:
+#   1) latest tag to match
+#   2) latest tag alias to apply if tag matches latest tag 
+set -eo pipefail
+set -x
+source rapids-constants
+export RAPIDS_SCRIPT_NAME="rapids-docker-multiarch-from-s3"
+
+latest_tag="$1:-"
+latest_tag_alias="$2:-"
+
+DOCKER_TMP_DIR="$(mktemp -d)"
+
+DOCKER_STARTS_WITH="docker_"
+DOCKER_ENDS_WITH=".tar.gz"
+
+S3_PATH=$(rapids-s3-path)
+BUCKET_PREFIX=${S3_PATH/s3:\/\/${RAPIDS_DOWNLOADS_BUCKET}\//} # removes s3://rapids-downloads/ from s3://rapids-downloads/ci/rmm/...
+
+# shellcheck disable=SC2016
+DOCKER_TARBALLS=$(
+  set -eo pipefail;
+  aws \
+    --output json \
+    s3api list-objects \
+    --bucket "${RAPIDS_DOWNLOADS_BUCKET}" \
+    --prefix "${BUCKET_PREFIX}" \
+    --page-size 100 \
+    --query "Contents[?contains(Key, '${DOCKER_STARTS_WITH}') && ends_with(Key, '${DOCKER_ENDS_WITH}')].Key" \
+    | jq -c
+)
+export DOCKER_TARBALLS
+
+# create an associative array (i.e. dict) of multiarch image to per-arch image
+declare -A MULTIARCH_IMAGES
+
+# download and load them all
+for OBJ in $(jq -nr 'env.DOCKER_TARBALLS | fromjson | .[]'); do
+  FILENAME=$(basename "${OBJ}")
+  S3_URI="${S3_PATH}${FILENAME}"
+
+  rapids-echo-stderr "Downloading ${S3_URI} into ${DOCKER_TMP_DIR}"
+  aws s3 cp --only-show-errors "${S3_URI}" "${DOCKER_TMP_DIR}"/ 1>&2
+
+  rapids-echo-stderr "Loading into docker"
+  loaded_image=$(docker load < "${DOCKER_TMP_DIR}/${FILENAME}")
+  loaded_image="${loaded_image/"Loaded image: "/}"
+
+  # delete the tarball to save space
+  rm "${DOCKER_TMP_DIR}/${FILENAME}" 1>&2
+
+  # strip -$(uname -m) or amd64 or arm64
+  loaded_image_no_arch="${loaded_image/"-x86_64"/}"
+  loaded_image_no_arch="${loaded_image_no_arch/"-aarch64"/}"
+  loaded_image_no_arch="${loaded_image_no_arch/"-amd64"/}"
+  loaded_image_no_arch="${loaded_image_no_arch/"-arm64"/}"
+
+  # no-arch tag is the final multiarch tag
+  multiarch_tag="${loaded_image_no_arch}"
+
+  # store per-arch image in the associative array by multiarch tag
+  MULTIARCH_IMAGES["${multiarch_tag}"]+=" ${loaded_image}"
+done
+
+manifests_to_copy=()
+
+for key in "${!MULTIARCH_IMAGES[@]}"; do
+  values="${MULTIARCH_IMAGES[$key]}"
+  rapids-echo-stderr "Preparing multiarch manifest for: ${key} with per-arch images: ${values}"
+
+  manifest_args=""
+  for value in ${values}; do
+    # use the local registry set up by the `service` block in GHA
+    local_name="localhost:5000/${value}"
```

Comment:
Where does the port number come from? Could we avoid hardcoding it here? If we have to hardcode it in the script, I see it used in multiple places so at least a single variable would be helpful.

Reply:
It's the default port of the docker registry image (launched in the containing workflow). Agreed that it makes less sense for this script to contain a hardcoded value that only makes sense in the outer workflow.

If this were to be transformed into a variable, it could be either an argument or env var, something like `INTERMEDIATE_DOCKER_REGISTRY` or `LOCAL_DOCKER_REGISTRY`?

Reply:
I like `LOCAL_DOCKER_REGISTRY`.

Reply:
Ideally with a comment indicating where this requirement is coming from.